### ISSUE TITLE
bam2fastq doesn't give back the original fastq files

### ISSUE THREAD

**Issue by eschmauch** [nunofonseca/fastq_utils#12]
Hi ! 
I'm processing publicly available data in the forms of bam files that originate from the fastq2bam command. 
I'm thus trying to retrieve the original fastq files using the bam2fastq command. 
I'm not very much familiar with what that command does, but it seems that the fastq files outputted by bam2fastq have shorter reads than the ones inputted to fastq2bam when converting them to bam files. They are thus too short for processing with cellranger. 

I might be using the package in a wrong way but I believed that fastq2bam is used for lossless conversion of fastq files to bam files and then bam2fastq to retrieve the fastq files. 

Thank you in advance for your help.

Best, 

Eloi

**Comment by nunofonseca** [comment]
Hi!

If the BAM file provided to bam2fastq has single cell data (e.g., cdna, umi, cell barcode, ...) then bam2fastq will create multiple fastq files with the same prefix: one file with the cdna sequences, another with the respective umi, cell barcodes, etc. 

It is possible to add the feature to recreate the fastq files with the umi, cell barcodes, etc in the original 'location' of the read. Would this feature be useful to you? If so, can you please provide an example of a BAM file (just the URL to the repository) that you are trying to convert?

Cheers,
Nuno

**Comment by pengpengyang94** [comment]
hi, 
I also encountered the same problem.The linked file will lose the 12bp UMI sequence. 3’ gene expression libraries are prepared
according to the manufacturer’s instructions of the v2 or v3 Chromium Single Cell Reagent Kits (10X Genomics).
http://ftp.sra.ebi.ac.uk/vol1/run/ERR437/ERR4375167/
Thank you in advance for your help.

**Comment by jlduan** [comment]
I have the same problem. The EBI downloaded raw bam files are not really 'untouched' from sequencer. The paper did not really describe the processing of the uploaded data. After some digging, I think the bam files I downloaded are processed by fastq_utils.

They are 10x genomics' libraries (one example: ftp.sra.ebi.ac.uk/vol1/run/ERR437/ERR4375413/5f1b1699fe9c934c8b83624d_1.bam). 

I have to guess the meaning of the tags in the bam.
on: read name
op: read 1 quality
CR: cell barcode
CY: cell barcode quality
QX: UMI
OQ: UMI quality
BC: index
QT: index quality

I don't really understand the rationale why not just uploading raw fastq files. If the purpose is storage efficiency, then why the op tag is included (redundant? ), since the same line already has sequencing qualities of cell barcode (CY), UMI (QX).

The name of the tool is called 'fastq_utils', which kind of demonstrating the importance of fastq format, it would be nice that it can convert processed bams to original fastq files.

Maybe I am miss something, is there some documentations/papers describing the reasoning, or some paper that I can cite when using it.

Thanks!

**Comment by Wkt8** [comment]
Hi! The raw fastq files are also available at the HCA Data Coordination Platform, for the datasets with raw fastq files available. They can be found here [here](https://data.humancellatlas.org/explore/files?catalog=dcp2&filter=%5B%7B%22facetName%22:%22projectTitle%22,%22terms%22:%5B%22Cells%20of%20the%20adult%20human%20heart%22%5D%7D%5D) for that specific project.

**Comment by jlduan** [comment]
@Wkt8 

That HCA website is not very useful. Their curl link is only valid for one hour and  there is no way to manually select individual files.

**Comment by nunofonseca** [comment]
Thanks for the feedback and for the links. The possibility to (re)create the fastq files from the BAM is clearly a wanted feature that I will add in the coming weeks.

The tags used in the BAM files follow the SAM/BAM format specification (https://samtools.github.io/hts-specs/SAMv1.pdf) with the exception of the  'on' and 'op' tags.  

Cheers.

**Comment by ESapenaVentura** [comment]
Hi @jlduan - For your query, there are 2 options:
- Using the filters above, you can select the data for your requisites (e.g. In [this example](https://data.humancellatlas.org/explore/files?catalog=dcp2&filter=%5B%7B%22facetName%22:%22projectTitle%22,%22terms%22:%5B%22Cells%20of%20the%20adult%20human%20heart%22%5D%7D,%7B%22facetName%22:%22organismAgeRange%22,%22terms%22:%5B%7B%22ageMax%22:1734480000,%22ageMin%22:1576800000,%22ageUnit%22:%22year%22%7D%5D%7D%5D), I have filtered for between 50 and 55 years old). If you click export selected data from there, you can get a curl command to only download that data.
- If you want more flexibility on the search, the [data browser API - get index files endpoint](https://service.azul.data.humancellatlas.org/#/Index/get_index_files) provides you with the ability to perform an ElasticSearch query. For each hit of the query, there is a "files" field that will provide you with an URL. Following that URL will provide you with a webpage [similar to this one](https://service.azul.data.humancellatlas.org/fetch/repository/files/5772511f-2bae-4948-ae79-ff2b3b004aa7?version=2019-09-20T10%3A40%3A09.415146Z&catalog=dcp2), and a GET request to the "Location" URL will provide you with the desired file(s)

I hope this can help you!

**Comment by hannes-ucsc** [comment]
> provides you with the ability to perform an ElasticSearch query

Not quite. It's a proprietary filter syntax, not nearly as flexible as ES, but also much simpler.

@jdjuan

Note the `hits[].files[].url` property of the response that @ESapenaVentura pointed out. That URL itself does not expire, it redirects to one that does.

> That HCA website is not very useful. Their curl link is only valid for one hour and there is no way to manually select individual files.

The link to the manifest should be valid for a week. I've opened https://github.com/DataBiosphere/azul/issues/2875 to pursue this.  The individual URLs in the manifest don't expire, but just like the URLs in the `/index/files` response above, the URL they redirect to are valid for an hour. The redirect is followed immediately by `curl` so the expiration should only get in the way for files that take more than one hour to download. I'd like to learn more about your experience with the curl download from the HCA Data Browser and how it was possible that you ran into the one hour expiration. Our intention is to fine tune the expiration of URL so it does not get in the way, without sacrificing security and privacy.

**Comment by jlduan** [comment]
Hi 

@ESapenaVentura

Thank you very much for the reply.

I know I can filter individual samples. But sometimes, some downloaded files are corrupted after sha256 checking, it is not very straightforward to just re-download selected files. And in most of my cases, I don’t need the read index (I1) fastq files, it seems that I have to download them with R1 and R2 altogether.

I am not familiar with the data brower API, I will look into it. Thank you!



@hannes-ucsc

Thank you very much for the additional explanation. Appreciate.
The info/warning I got when I requested curl is that it would remain valid for one hour. 

<img width="646" alt="Screen Shot 2021-03-08 at 14 20 02" src="https://user-images.githubusercontent.com/699489/110378856-bbe8ca80-801b-11eb-9024-5c891ed4149d.png">
